syncstorage-spanner/src/manager/session.rs

```diff
@@ -21,25 +22,103 @@ pub struct SpannerSession {
     pub session: Session,
     /// The underlying client (Connection/Channel) for interacting with spanner
     pub client: SpannerClient,
-    pub(crate) use_test_transactions: bool,
+    /// Session Settings
+    pub settings: SpannerSessionSettings,
     /// A second based UTC for SpannerSession creation.
     /// Session has a similar `create_time` value that is managed by protobuf,
     /// but some clock skew issues are possible.
     pub(crate) create_time: i64,
-    /// Whether we are using the Spanner emulator
-    pub using_spanner_emulator: bool,
+}
+
+impl SpannerSession {
+    /// Return [CallOption] including a Dynamic Routing header for this
+    /// [SpannerSession] in its [grpcio::Metdata]
+    ///
+    /// The more common form of [grpcio::Metdata] used by Spanner operations
+    /// such as `begin`, `commit`, `execute_sql`, etc.
+    pub fn session_opt(&self) -> Result<CallOption, grpcio::Error> {
+        // NOTE: this could also be cached (then cloned)
+        let meta = self
+            .settings
+            .metadata_builder()
+            .routing_param("session", self.session.get_name())
+            .build()?;
+        Ok(CallOption::default().headers(meta))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct SpannerSessionSettings {
+    /// The database name
+    pub database: String,
+
+    /// Whether [SpannerDb] uses mutations, which should be more efficient for
+    /// some of its bulk operations
+    pub use_mutations: bool,
+
+    /// Whether the Leader Aware Routing header should be included in gRPC
+    /// metdata
+    pub route_to_leader: bool,
+
+    /// Max age of a Session
+    pub max_lifespan: Option<u32>,
+    /// Max idle time of a Session
+    pub max_idle: Option<u32>,
+
+    /// For tests: disables transactions from committing
+    pub(crate) use_test_transactions: bool,
+    /// Spanner emulator hostname when set to Spanner emulator mode
+    pub emulator_host: Option<String>,
+}
+
+impl SpannerSessionSettings {
+    pub fn from(settings: &Settings) -> Result<Self, DbError> {
```

Comment:
Could this be a `TryFrom`?

Reply:
Last time we discussed this I came to the conclusion that I'd rather leave the `Try`/`TryFrom` for simpler type conversions that feel a bit more implicit (like `let foo: i64 = 1i32.into();`) and be more explicit for things that are more along the lines of construction or deserialization (like `serde_json::from_str`);

Renamed it `from` -> `from_settings` like we have elsewhere to distinguish it.

Reply:
or `str::parse` is maybe a better example